cdap-ui/app/cdap/components/Replicator/Create/Content/TargetConfig/TargetList/index.tsx

```diff
@@ -0,0 +1,163 @@
+import classnames from 'classnames';
+
+const styles = (theme): StyleRules => {
+  return {
+    header: {
+      display: 'grid',
+      gridTemplateColumns: '1fr 400px',
+    },
+    searchSection: {
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'flex-end',
+    },
+    targetItem: {
+      marginRight: '25px',
+      cursor: 'pointer',
+    },
+    selected: {
+      backgroundColor: theme.palette.grey[700],
+    },
+    search: {
+      width: '200px',
+      marginLeft: '20px',
+    },
+    listContainer: {
+      marginTop: '15px',
+      marginBottom: '15px',
+    },
+  };
+};
+
+interface ITargetListProps extends WithStyles<typeof styles> {
+  onSelect: (target) => void;
+  currentSelection: any;
+}
+
+const TargetListView: React.FC<ITargetListProps> = ({ classes, onSelect, currentSelection }) => {
+  const [targets, setTargets] = React.useState([]);
+  const [widgetMap, setWidgetMap] = React.useState({});
+  const [search, setSearch] = React.useState('');
+
+  React.useEffect(() => {
+    fetchPluginsAndWidgets(PluginType.target).subscribe((res) => {
+      if (res.plugins.length === 1 && !currentSelection) {
+        onSelect(res.plugins[0]);
+      }
+
+      setTargets(res.plugins);
+      setWidgetMap(res.widgetMap);
+    });
+  }, []);
+
+  function handleSearch(e) {
+    const value = e.target.value;
+    setSearch(value);
+  }
+
+  if (targets.length === 1) {
+    return null;
+  }
+
+  let filteredTarget = targets;
+  if (search.length > 0) {
+    filteredTarget = targets.filter((target) => {
+      const pluginKey = `${target.name}-${target.type}`;
+      const displayName = objectQuery(widgetMap, pluginKey, 'display-name');
+
+      const normalizedSearch = search.toLowerCase();
+
+      if (target.name.toLowerCase().indexOf(normalizedSearch) !== -1) {
+        return true;
+      }
+
+      return displayName && displayName.toLowerCase().indexOf(normalizedSearch) !== -1;
+    });
+  }
+
+  return (
+    <div>
+      <div className={classes.header}>
+        <div>
+          <h4>Select target</h4>
+        </div>
+        <div className={classes.searchSection}>
+          <div>
+            {targets.length} {targets.length === 1 ? 'source' : 'sources'} available
+          </div>
+          <TextField
+            className={classes.search}
+            value={search}
+            onChange={handleSearch}
+            size="small"
+            variant="outlined"
+            placeholder="Search source by name"
+            InputProps={{
+              startAdornment: (
+                <InputAdornment position="start">
+                  <Search />
+                </InputAdornment>
+              ),
+            }}
+          />
+        </div>
+      </div>
+
+      <div className={classes.listContainer}>
+        <HorizontalCarousel scrollAmount={150}>
+          {filteredTarget.map((target) => {
+            const pluginKey = `${target.name}-${target.type}`;
```

Comment:
I only see this twice, but might be nice to have a helper function for generating the plugin keys

Reply:
I don't think this is necessary. In this particular case it just so happen that it's the same, but it could very well be different depending on whether later on we will add different things to the source vs target (such as driver version info, etc.)